### PR TITLE
Support cxx extensions

### PR DESCRIPTION
**Pull request by nirvdrum**

### CODE REVIEW

File: lib/cext/ruby/ruby.h [hunk 1]
```diff
@@ -1044,7 +1044,7 @@ struct RFile {
 #define RCOMPLEX_SET_IMAG(cmp, i) RB_OBJ_WRITE((cmp), &((struct RComplex *)(cmp))->imag,(i))
 
 struct RData {
-    struct RBasic basic;
+    //struct RBasic basic;  // RBasic is an empty struct. clang makes it size 0 for C, but size 1 for C++. That difference affects field offsets, so we comment out the reference to ensure the size is always 0.
```
Comment (eregon) [inline lib/cext/ruby/ruby.h]: Maybe add a `Truffle:` prefix here.

Comment (nirvdrum) [reply lib/cext/ruby/ruby.h]: Done. f2c9004b90ef2b88469c340b352fd33fb31d16b8

File: lib/cext/ruby/ruby.h [hunk 2]
```diff
@@ -1502,7 +1502,7 @@ rb_ulong2num_inline(unsigned long v)
     if (RB_POSFIXABLE(v))
 	return LONG2FIX(v);
     else
-	return rb_uint2big(v);
+	return truffle_invoke(RUBY_CEXT, "rb_ulong2num", (long) v);
```
Comment (eregon) [inline lib/cext/ruby/ruby.h]: I can't see where rb_ulong2num is defined, did you mean `ULONG2NUM`?

File: test/truffle/cexts/unf_ext/unf_ext.sh [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
+source test/truffle/common.sh.inc
+
+jt gem-test-pack
+
+export TRUFFLERUBY_CEXT_ENABLED=true
+jt ruby -S gem install --local "$(jt gem-test-pack)/gem-cache/unf_ext-0.0.7.4.gem" -V -N --backtrace
```
Comment (eregon) [inline test/truffle/cexts/unf_ext/unf_ext.sh]: :+1: This is much more useful than tweaking our internal cextc/compile_cext :)

Comment (nirvdrum) [reply test/truffle/cexts/unf_ext/unf_ext.sh]: Well, I just updated jt.rb, which already has a whole bunch of gem-specific hacks in it ;-)

File: lib/cext/ruby/ruby.h (hunk 2 above)
Comment (nirvdrum) [reply lib/cext/ruby/ruby.h]: @bjfish This was your suggestion. Any ideas?

Comment (eregon) [reply lib/cext/ruby/ruby.h]:
```
$ jt ruby -e 'p Truffle::CExt.rb_ulong2num'
-e:1:in `method_missing': undefined method `rb_ulong2num' for Truffle::CExt:Module (NoMethodError)
Did you mean?  rb_long2int
               rb_num2ulong
               rb_num2long
               ULONG2NUM
	from -e:1:in `<main>'
```

Comment (bjfish) [reply lib/cext/ruby/ruby.h]: In `cext.rb` we can add:
```
def rb_ulong2num
   raise NotImplementedError, 'rb_ulong2num not implemented'
end
```

Comment (eregon) [reply lib/cext/ruby/ruby.h]: That's less confusing than a NoMethodError, +1

File: test/truffle/cexts/unf_ext/unf_ext.sh (hunk 3 above)
Comment (eregon) [reply test/truffle/cexts/unf_ext/unf_ext.sh]: And less hacks is better than more hacks :)

**Pull request merged**